File: 452_MinimumNumberOfArrowsToBurstBalloons.cpp
//28.94%
class Solution {
public:
    int findMinArrowShots(vector<pair<int, int>>& points) {
        if(points.empty())
			return 0;
		
		auto cmp = [](const pair<int, int>& p1, const pair<int, int>& p2){ return p1.first < p2.first; };
		sort(points.begin(), points.end(), cmp);
		int cnt(1);
		for(int i = 1; i < points.size(); ++i){
			if(points[i].first <= points[i - 1].second)
				points[i].second = min(points[i - 1].second, points[i].second);
			else
				++cnt;
		}
		return cnt;
    }
};

int findMinArrowShots(vector<pair<int, int>>& points) {
        int count = 0, arrow = INT_MIN;
        sort(points.begin(), points.end(), mysort);
        for(int i = 0; i<points.size(); i++){
            if(points[i].first > arrow){
				arrow = points[i].second; // new arrow shot the end of points[i]
				++count;
			}
        }
        return count;
    }
    static bool mysort(const pair<int, int>& a, const pair<int, int>& b){
        return a.second<b.second;
    }

File: 452_MinimumNumberOfArrowsToBurstBalloons.py
#19.81%
class Solution(object):
    def findMinArrowShots(self, points):
        """
        :type points: List[List[int]]
        :rtype: int
        """
        if not points:
            return 0
        cnt = 1
        points.sort(key = lambda point: point[0])
        for i in xrange(1, len(points)):
            if points[i][0] <= points[i - 1][1]:
                points[i][1] = min(points[i - 1][1], points[i][1])
            else:
                cnt += 1
        return cnt

#69.56%		
class Solution(object):
    def findMinArrowShots(self, points):
        """
        :type points: List[List[int]]
        :rtype: int
        """
        points = sorted(points, key = lambda x: x[1])
        res, end = 0, -float('inf')
        for interval in points:
            if interval[0] > end:
                res += 1
                end = interval[1]
        return res